### PR TITLE
(SDK-276) re-do rubocop configuration

### PR DESCRIPTION
**Pull request by DavidS**
This uses hardcoded rubocop_defaults to build a minimal .rubocop.yml
config. The configuration is minimal (only list what you want), and can
be easily overridden.

The profile_builder.rb script can help in rebuilding the rubocop_defaults
file for new rubocop versions.

The cleanup_cops.yml contains a list of basic cops everyone should check.
This was originally derived from the main PDK rubocop configuration.
See https://github.com/puppetlabs/pdk/pull/78.

Downstream users can choose from four different strictness
profiles: off, cleanups_only, strict, hardcore.

* off: no cops enabled, use at your own risk.
* cleanups_only: a set of uncontroversial cops that will cleanup your
  codebase nicely. All of them have autocorrect enabled and are safe to
  apply.
* strict: the default. A good mix of cops following the established best
  practices of the community. The work required to pass this, will make
  your code base so much better.
* hardcore: enables all rubocop default cops. Still uses the recommended
  configs from strict, but enabled additional, unvetted tests. Have fun!

### CODE REVIEW

File: moduleroot/.rubocop.yml.erb [hunk 1]
```diff
@@ -1,26 +1,37 @@
...
+path = File.absolute_path(File.join(__FILE__, '..', '..', "rubocop_defaults-#{RuboCop::Version.version}.yml"))
+default_enabled_cops = YAML.load(File.read(path))[:default_enabled_cops]
+
+enabled_cops = profile['enabled_cops'] || {}
+enabled_cops = default_enabled_cops if enabled_cops == 'all'
+(enabled_cops - default_enabled_cops).sort.each { |c| configs[c] ||= {}; configs[c]['Enabled'] = true }
+(default_enabled_cops - enabled_cops).sort.each { |c| configs[c] ||= {}; configs[c]['Enabled'] = false }
+-%>
+<%=
+require 'yaml'
```
Comment (rodjek) [inline moduleroot/.rubocop.yml.erb]: Is this require necessary (unconditionally required in the previous block)?

File: moduleroot/.rubocop.yml.erb [hunk 2]
```diff
@@ -1,26 +1,37 @@
...
-<%- end -%>
+require 'deep_merge/core'
+profile = @configs['profiles'][@configs['selected_profile']]
+configs = defaults.dup
+DeepMerge::deep_merge!(profile['configs'], configs, knockout_prefix: "--", preserve_unmergeables: false)
+
+require 'yaml'
+require 'rubocop/version'
+path = File.absolute_path(File.join(__FILE__, '..', '..', "rubocop_defaults-#{RuboCop::Version.version}.yml"))
+default_enabled_cops = YAML.load(File.read(path))[:default_enabled_cops]
```
Comment (rodjek) [inline moduleroot/.rubocop.yml.erb]: Should probably use `YAML.safe_load` here (same for other uses in `profile_builder.rb`)

File: profile_builder.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+default_configs = YAML.load(`rubocop --show-cops --require rubocop-rspec`)
+all_cops = default_configs.keys - [ 'AllCops', 'require' ]
+default_enabled_cops = all_cops.find_all { |c| default_configs[c]['Enabled'] }
+default_disabled_cops = all_cops - default_enabled_cops
+
+$stderr.puts "Found #{default_enabled_cops.length} enabled, and #{default_disabled_cops.length} disabled cops in the default config."
+
+# fetch config from current PDK
+FileUtils.cp("../pdk/.rubocop.yml", '.')
+File.open('.rubocop_todo.yml', 'w') do |f|
```
Comment (rodjek) [inline profile_builder.rb]: Why's this file being opened?

Comment (DavidS) [inline profile_builder.rb]: it is included in the pdk config, and rubocop complains unless it's there.

File: moduleroot/.rubocop.yml.erb (hunk 1 above)
Comment (DavidS) [inline moduleroot/.rubocop.yml.erb]: fixed

File: moduleroot/.rubocop.yml.erb (hunk 2 above)
Comment (DavidS) [inline moduleroot/.rubocop.yml.erb]: the default rubocop config contains Symbols and other crud that won't load under safe_load. :-(

**Pull request merged**